client/state/media/reducer.js

```diff
@@ -2,22 +2,156 @@
  * External dependencies
  */
 
-import { omit } from 'lodash';
+import { isEmpty, mapValues, omit, pickBy, without } from 'lodash';
 /**
  * Internal dependencies
  */
 import {
 	MEDIA_DELETE,
+	MEDIA_ERRORS_CLEAR,
+	MEDIA_ITEM_ERRORS_CLEAR,
+	MEDIA_ITEM_CREATE,
 	MEDIA_ITEM_REQUEST_FAILURE,
 	MEDIA_ITEM_REQUEST_SUCCESS,
 	MEDIA_ITEM_REQUESTING,
 	MEDIA_RECEIVE,
 	MEDIA_REQUEST_FAILURE,
 	MEDIA_REQUEST_SUCCESS,
 	MEDIA_REQUESTING,
+	MEDIA_SOURCE_CHANGE,
 } from 'state/action-types';
 import { combineReducers, withoutPersistence } from 'state/utils';
 import MediaQueryManager from 'lib/query-manager/media';
+import { validateMediaItem } from 'lib/media/utils';
+import { ValidationErrors as MediaValidationErrors } from 'lib/media/constants';
+
+const isExternalMediaError = message =>
+	message.error && ( message.error === 'servicefail' || message.error === 'keyring_token_error' );
+
+const isMediaError = action =>
+	action.error && ( action.siteId || isExternalMediaError( action.error ) );
+
+/**
+ * Returns the updated media errors state after an action has been
+ * dispatched. The state reflects a mapping of site ID, media ID pairing to
+ * an array of errors that occurred for that corresponding media item.
+ *
+ * @param  {object} state  Current state
+ * @param  {object} action Action payload
+ * @returns {object}        Updated state
+ */
+export const errors = ( state = {}, action ) => {
+	switch ( action.type ) {
+		case MEDIA_ITEM_CREATE: {
+			if ( ! action.site || ! action.transientMedia ) {
+				return state;
+			}
+
+			const items = Array.isArray( action.transientMedia )
+				? action.transientMedia
+				: [ action.transientMedia ];
+			const mediaErrors = items.reduce( function( memo, item ) {
+				const itemErrors = validateMediaItem( action.site, item );
+				if ( itemErrors.length ) {
+					memo[ item.ID ] = itemErrors;
+				}
+
+				return memo;
+			}, {} );
+
+			return {
+				...state,
+				[ action.site.ID ]: {
+					...state[ action.site.ID ],
+					...mediaErrors,
+				},
+			};
+		}
+
+		case MEDIA_ITEM_REQUEST_FAILURE:
+		case MEDIA_REQUEST_FAILURE: {
+			// Track any errors which occurred during upload or getting external media
+			if ( ! isMediaError( action ) ) {
+				return state;
+			}
+
+			let mediaErrors;
+
+			if ( Array.isArray( action.error.errors ) ) {
```

Comment:
want to consolidate down onto the `? :` setter here like up on line 50?

Reply:
Good call, done in https://github.com/Automattic/wp-calypso/pull/39375/commits/b70775af792fb82577edc82e2aebd52707bb7ff6